[Bug]: Not allowed javaletter or keyword in identifier found

### Describe the bug

Hello, I have this error, even with the latest version. Could you tell me how I can debug the execution to see which files cause the issue? Thanks.

```
[ERROR] Unexpected error
spoon.JLSViolation: Not allowed javaletter or keyword in identifier found. See JLS for correct identifier. Identifier: K'
    at spoon.JLSViolation.throwIfSyntaxErrorsAreNotIgnored (JLSViolation.java:38)
```

### Source code you are trying to analyze/transform

_No response_

### Source code for your Spoon processing

_No response_

### Actual output

_No response_

### Expected output

_No response_

### Spoon Version

10.4.2

### JVM Version

openjdk 17.0.7 2023-04-18 LTS OpenJDK Runtime Environment Zulu17.42+19-CA (build 17.0.7+7-LTS) OpenJDK 64-Bit Server VM Zulu17.42+19-CA (build 17.0.7+7-LTS, mixed mode, sharing)

### What operating system are you using?

macOS

--- a/src/main/java/spoon/support/compiler/jdt/ReferenceBuilder.java
+++ b/src/main/java/spoon/support/compiler/jdt/ReferenceBuilder.java
@@ -58,6 +58,7 @@ import org.eclipse.jdt.internal.compiler.lookup.RawTypeBinding;
 import org.eclipse.jdt.internal.compiler.lookup.ReferenceBinding;
 import org.eclipse.jdt.internal.compiler.lookup.Scope;
 import org.eclipse.jdt.internal.compiler.lookup.SourceTypeBinding;
+import org.eclipse.jdt.internal.compiler.lookup.SyntheticFactoryMethodBinding;
 import org.eclipse.jdt.internal.compiler.lookup.TypeBinding;
 import org.eclipse.jdt.internal.compiler.lookup.TypeVariableBinding;
 import org.eclipse.jdt.internal.compiler.lookup.UnresolvedReferenceBinding;
@@ -422,7 +423,11 @@ public class ReferenceBuilder {
 		if (sourceStart >= 0 && sourceEnd >= 0) {
 			ref.setPosition(jdtTreeBuilder.getPositionBuilder().buildPosition(sourceStart, sourceEnd));
 		}
-		if (exec.isConstructor()) {
+		// JDT creates synthetic <factory> methods for inference of diamond constructors (We guess they had that
+		// lying around). If the type can not be completely resolved, e.g. due to no classpath, this factory is
+		// not replaced and appears in the AST. We need to fix its name so e.g. `CtExecutableReference#isConstructor`
+		// works.
+		if (exec.isConstructor() || exec.original() instanceof SyntheticFactoryMethodBinding) {
 			ref.setSimpleName(CtExecutableReference.CONSTRUCTOR_NAME);
 
 			// in case of constructor of an array, it's the return type that we want
@@ -1023,7 +1028,15 @@ public class ReferenceBuilder {
 			return getTypeReferenceOfBoundingType(binding).clone();
 		} else {
 			CtTypeReference<?> ref = this.jdtTreeBuilder.getFactory().Core().createTypeParameterReference();
-			ref.setSimpleName(new String(binding.sourceName()));
+			String name = new String(binding.sourceName());
+			if (binding.declaringElement instanceof SyntheticFactoryMethodBinding) {
+				// JDT uses these factory methods for type inference of diamond constructors. In no classpath mode they
+				// might be left around. They append a variable number of primes (') to the original name, which is not
+				// valid in Java. We undo this here and hope for the best.
+				name = name.replace("'", "");
+			}
+
+			ref.setSimpleName(name);
 			return ref;
 		}
 	}

--- a/src/test/java/spoon/test/api/NoClasspathTest.java
+++ b/src/test/java/spoon/test/api/NoClasspathTest.java
@@ -23,11 +23,14 @@ import java.util.List;
 import org.junit.jupiter.api.Test;
 
 import spoon.Launcher;
+import spoon.reflect.CtModel;
+import spoon.reflect.code.CtConstructorCall;
 import spoon.reflect.code.CtFieldAccess;
 import spoon.reflect.code.CtInvocation;
 import spoon.reflect.code.CtLocalVariable;
 import spoon.reflect.code.CtReturn;
 import spoon.reflect.declaration.CtClass;
+import spoon.reflect.declaration.CtElement;
 import spoon.reflect.declaration.CtField;
 import spoon.reflect.declaration.CtMethod;
 import spoon.reflect.declaration.CtType;
@@ -38,7 +41,11 @@ import spoon.reflect.visitor.filter.TypeFilter;
 import spoon.support.SpoonClassNotFoundException;
 import spoon.support.visitor.SignaturePrinter;
 import spoon.test.api.testclasses.Bar;
+import spoon.testing.assertions.SpoonAssertions;
+import spoon.testing.utils.ModelTest;
 
+import static java.util.function.Predicate.not;
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 import static org.junit.jupiter.api.Assertions.assertNull;
@@ -200,5 +207,20 @@ public class NoClasspathTest {
 		CtTypeReference myInterfaceReference = spoon.getFactory().Type().createReference("org.myorganization.MyInterface");
 		assertTrue(myClassReference.isSubtypeOf(myInterfaceReference));
 		assertTrue(field.getType().isSubtypeOf(myInterfaceReference));
+	}
+
+	@ModelTest("src/test/resources/noclasspath/issue5591/DiamondConstructorCallTypeInference.java")
+	void testJdtFactoryMethodsForDiamond(CtModel model) {
+		// contract: Leftover <factory> methods from JDT's diamond constructor type inference are handled in
+		// no classpath mode
+		var constructorCalls = model.getElements(new TypeFilter<>(CtConstructorCall.class));
+		assertThat(constructorCalls).hasSize(1);
+
+		var executableReference = constructorCalls.get(0).getExecutable();
+		SpoonAssertions.assertThat(executableReference)
+			.isNotNull()
+			.nested(it -> it.getSimpleName().isEqualTo("<init>"))
+			.nested(it -> it.satisfies(CtExecutableReference::isConstructor))
+			.nested(it -> it.getParameters().hasSize(1));
 	}
 }

--- a/src/test/java/spoon/testing/assertions/SpoonAssert.java
+++ b/src/test/java/spoon/testing/assertions/SpoonAssert.java
@@ -1,10 +1,17 @@
 package spoon.testing.assertions;
 
+import java.util.function.Consumer;
+
 public interface SpoonAssert<SELF, ACTUAL> {
 
 	SELF self();
 
 	ACTUAL actual();
+
+	default SELF nested(Consumer<SELF> checks) {
+		checks.accept(self());
+		return self();
+	}
 
 	void failWithMessage(String errorMessage, Object... arguments);
 }

--- a/src/test/resources/noclasspath/issue5591/DiamondConstructorCallTypeInference.java
+++ b/src/test/resources/noclasspath/issue5591/DiamondConstructorCallTypeInference.java
@@ -1,0 +1,14 @@
+package com.foo.bar;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+public class DiamondConstructorCallTypeInference.java {
+	// Whatever does not exist.
+	private final List<? extends Whatever> items;
+
+	public DiamondConstructorCallTypeInference.java(Collection<? extends Whatever> items) {
+		this.items = new ArrayList<>(items);
+	}
+}

